File: 2585-number-of-ways-to-earn-points/2585-number-of-ways-to-earn-points.py
class Solution:
    def waysToReachTarget(self, target: int, types: List[List[int]]) -> int:
            MOD = 10**9 + 7
            dp = [0] * (target + 1)
            dp[0] = 1

            for count, marks in types:
                for j in range(target, -1, -1):
                    for k in range(1, count + 1):
                        if j - k * marks < 0:
                            break
                        dp[j] = (dp[j] + dp[j - k * marks]) % MOD

            return dp[target]

File: 2585-number-of-ways-to-earn-points/2585-number-of-ways-to-earn-points.cpp
class Solution {
private:
    int mod = 1e9 + 7;
    
    int _countSubsetSum(int target, vector<vector<int>>& types, int n, vector<vector<int>>& store) {
        if (target == 0) {
            return 1;
        }
        
        if (n == 0) {
            return 0;
        }
        
        if (store[n][target] != -1) {
            return store[n][target];
        }
        
        int questionCount = types[n - 1][0];
        int questionMarks = types[n - 1][1];
        int result = 0;
        for (int i = 0; i <= questionCount; i++) {
            int totalMarks = i * questionMarks;
            
            if (totalMarks <= target) {
                result = (result + _countSubsetSum(target - totalMarks, types, n - 1, store)) % mod;
            }
        }
        
        store[n][target] = result;
        
        return store[n][target];
    }
    
public:
    int waysToReachTarget(int target, vector<vector<int>>& types) {        
        int n = types.size();
        vector<vector<int>> store(n + 1, vector<int>(target + 1, -1));
        
        return _countSubsetSum(target, types, n, store);
    }
};